planner/core/task.go

```diff
@@ -2374,15 +2386,17 @@ func (t *mppTask) convertToRootTaskImpl(ctx sessionctx.Context) *rootTask {
 	}.Init(ctx, t.p.SelectBlockOffset())
 	p.stats = t.p.statsInfo()
 	collectPartitionInfosFromMPPPlan(p, t.p)
+	rowSize := math.Max(collectRowSizeFromMPPPlan(t.p), 1) // use 1 as lower-bound for safety
 
-	cst := t.cst + t.count()*ctx.GetSessionVars().GetNetworkFactor(nil)
-	p.cost = cst / p.ctx.GetSessionVars().CopTiFlashConcurrencyFactor
+	cst := t.cst + t.count()*rowSize*ctx.GetSessionVars().GetNetworkFactor(nil)
+	cst /= p.ctx.GetSessionVars().CopTiFlashConcurrencyFactor
+	p.cost = cst
 	if p.ctx.GetSessionVars().IsMPPEnforced() {
-		p.cost = cst / 1000000000
+		cst /= 1000000000
```

Comment:
Should the `p.cost /= 1000000000` at the same time(like the previous)?

Reply:
I think we don't need to change `p.cost` here since `cst` is the one that used to be compared with other plan costs so we have to change it to make the planner choose this plan.
Keeping `p.cost` unchanged here can let us know the real cost of the plan when using `explain format=verbose`, and the new test `TestRowSizeInMPP` relies on this.